shared/src/types/keccak.rs

```diff
@@ -103,13 +104,48 @@ pub fn keccak_hash(bytes: &[u8]) -> KeccakHash {
 
 /// This module defines encoding methods compatible with Ethereum
 /// smart contracts.
+// TODO: relocate this sub-module to a new home
 pub mod encode {
     #[doc(inline)]
     pub use ethabi::token::Token;
     use tiny_keccak::{Hasher, Keccak};
 
     use super::*;
 
+    /// A container for data types that are able to be `encode`
+    /// or `encodePacked` Ethereum ABI-encoded.
+    #[derive(Clone, Debug, BorshSerialize, BorshDeserialize, BorshSchema)]
+    #[repr(transparent)]
+    pub struct EncodeCell<T> {
+        /// ABI-encoded value of type `T`.
+        encoded_data: Vec<u8>,
+        /// Indicate we do not own values of type `T`.
+        ///
+        /// Passing `PhantomData<T>` here would trigger the drop checker,
+        /// which is not the desired behavior, since we own an encoded value
+        /// of `T`, not a value of `T` itself.
+        _marker: PhantomData<*const T>,
+    }
+
+    impl<T> EncodeCell<T> {
```

Comment:
This is interesting. Should we change the return type of the `Encode::encode` method to return an instance of this struct? @sug0 ?

Reply:
I like it.